src/ophyd_async/core/_table.py

```diff
@@ -17,34 +22,31 @@ def _concat(value1, value2):
 class Table(BaseModel):
     """An abstraction of a Table of str to numpy array."""
 
-    model_config = ConfigDict(validate_assignment=True, strict=False)
-
-    @staticmethod
-    def row(cls: type[TableSubclass], **kwargs) -> TableSubclass:  # type: ignore
-        arrayified_kwargs = {}
-        for field_name, field_value in cls.model_fields.items():
-            value = kwargs.pop(field_name)
-            if field_value.default_factory is None:
-                raise ValueError(
-                    "`Table` models should have default factories for their "
-                    "mutable empty columns."
-                )
-            default_array = field_value.default_factory()
-            if isinstance(default_array, np.ndarray):
-                arrayified_kwargs[field_name] = np.array(
-                    [value], dtype=default_array.dtype
-                )
-            elif issubclass(type(value), Enum) and isinstance(value, str):
-                arrayified_kwargs[field_name] = [value]
+    # Allow extra so we can use this as a generic Table
+    model_config = ConfigDict(extra="allow")
+
+    @classmethod
+    def __init_subclass__(cls):
+        # But forbit extra in subclasses so it gets validated
+        cls.model_config = ConfigDict(validate_assignment=True, extra="forbid")
+        # Change fields to have the correct annotations
+        for k, anno in cls.__annotations__.items():
+            if get_origin(anno) is np.ndarray:
+                dtype = get_dtype(anno)
+                new_anno = Annotated[
+                    anno,
+                    NpArrayPydanticAnnotation.factory(
+                        data_type=dtype.type, dimensions=1, strict_data_typing=False
+                    ),
+                    Field(
+                        default_factory=lambda dtype=dtype: np.array([], dtype=dtype)
```

Comment:
Because it is created in a for loop, so needs to be bound to the function or it will disappear:
https://docs.astral.sh/ruff/rules/function-uses-loop-variable